File: src/main/java/com/github/ofrostdev/api/utils/database/DatabaseFactory.java
package com.github.ofrostdev.api.utils.database;

import com.henryfabio.sqlprovider.connector.SQLConnector;
import com.henryfabio.sqlprovider.connector.type.impl.MySQLDatabaseType;
import com.henryfabio.sqlprovider.connector.type.impl.SQLiteDatabaseType;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.plugin.Plugin;

import java.io.File;

public class DatabaseFactory {

    private final Plugin plugin;
    private final File dataFolder;

    public DatabaseFactory(Plugin plugin) {
        this.plugin = plugin;
        this.dataFolder = new File(plugin.getDataFolder(), "data");
    }

    public SQLConnector createConnector(ConfigurationSection section) {
        String databaseType = section.getString("type");
        ConfigurationSection typeSection = section.getConfigurationSection(databaseType);

        switch (databaseType) {
            case "sqlite":
                return buildSQLITE(typeSection).connect();
            case "mysql":
                return buildMYSQL(typeSection).connect();
            default:
                throw new UnsupportedOperationException("database type unsupported!");
        }
    }

    private SQLiteDatabaseType buildSQLITE(ConfigurationSection typeSection) {
        return SQLiteDatabaseType.builder()
                .file(new File(dataFolder, typeSection.getString("fileName")))
                .build();
    }

    private MySQLDatabaseType buildMYSQL(ConfigurationSection typeSection) {
        return MySQLDatabaseType.builder()
                .address(typeSection.getString("address"))
                .username(typeSection.getString("username"))
                .password(typeSection.getString("password"))
                .database(typeSection.getString("database"))
                .build();
    }

}

File: src/main/kotlin/com/github/ofrostdev/api/utils/storage/DatabaseFactory.kt
package com.github.ofrostdev.api.utils.storage

import com.henryfabio.sqlprovider.connector.SQLConnector
import com.henryfabio.sqlprovider.connector.type.impl.MySQLDatabaseType
import com.henryfabio.sqlprovider.connector.type.impl.SQLiteDatabaseType
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.plugin.Plugin
import java.io.File

class DatabaseFactory(private val plugin: Plugin) {
    private val dataFolder = File(plugin.dataFolder, "data")

    fun createConnector(section: ConfigurationSection): SQLConnector {
        val databaseType = section.getString("type")?.lowercase()
        val typeSection = section.getConfigurationSection(databaseType)

        return when (databaseType) {
            "sqlite" -> buildSQLITE(typeSection).connect()
            "mysql" -> buildMYSQL(typeSection).connect()
            else -> throw UnsupportedOperationException("database type unsupported!")
        }
    }

    private fun buildSQLITE(typeSection: ConfigurationSection): SQLiteDatabaseType {
        return SQLiteDatabaseType.builder()
            .file(File(dataFolder, typeSection.getString("fileName")))
            .build()
    }

    private fun buildMYSQL(typeSection: ConfigurationSection): MySQLDatabaseType {
        return MySQLDatabaseType.builder()
            .address(typeSection.getString("address"))
            .username(typeSection.getString("username"))
            .password(typeSection.getString("password"))
            .database(typeSection.getString("database"))
            .build()
    }

    companion object {
        fun getType(section: ConfigurationSection): String? {
            return section.getString("type")?.lowercase()
        }
    }
}
